BatchVAA protos and spy updates

justinschuldt opened the pull request:
Added nonce to the BatchID, and updated spy with a new endpoint that will return all types of VAAs, and allow filtering by `EmitterChain/TransactionID` or BatchID (`emitterChain/TransactionID/Nonce`)

jynnantonix commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+
+					// take an observation from the batch's list, set it to VAA
+					// so it's data can be considered for a filter match.
+					if len(b.Observations) > 0 {
```
Same comment here as above.  This is a nil-pointer dereference if the initial `vaa.Unmarshal` call is successful.

justinschuldt replied:
updated

jynnantonix commented on node/cmd/spy/spy.go:
```diff
@@ -185,10 +380,75 @@ func (s *spyServer) SubscribeSignedVAA(req *spyv1.SubscribeSignedVAARequest, res
+
+			case *spyv1.FilterEntry_EmitterFilter:
+				// validate the emitter address is valid by decoding it
+				_, err := decodeEmitterAddr(t.EmitterFilter.EmitterAddress)
```
Can we store the decoded address here rather than dropping it?  Then we don't need to decode it every time in `PublishSignedVAAByType`.

justinschuldt replied:
I agree, having the `spyv1.FlterEntry` type differ from the VAA's type for the same data is a mess. Previously, spy implemented a third set of structs that held transformed values of the `spyv1.EmitterFilter` values to be similar to the VAA's types ([`filter`](https://github.com/wormhole-foundation/wormhole/blob/b129fcb52dc531bdc7b8459cca9c9ec5d7e2863a/node/cmd/spy/spy.go#L81), [`subscription`](https://github.com/wormhole-foundation/wormhole/blob/b129fcb52dc531bdc7b8459cca9c9ec5d7e2863a/node/cmd/spy/spy.go#L86)) .

I don't think Spy should be in the business of transforming all the `spyv1.FilterEntry` values, handling the bad data, and then holding them in state for when a VAA comes along. So, I updated the proto types (all unused properties so OK to change), to align more closely with the `gossip` messages received by Spy. Now the filter structs of subscribers are held in state, and can be directly compared to the incoming `gossip` VAAs. This gets rid of the duplicate decoding, and a ton of boilerplate.

I also updated the response type of the new Spy endpoint so Spy can just forward the gossip messages along to subscribers without having to create custom response structs. I think the tight coupling is worth it here.

jynnantonix reviewed (commented):
Can we add some tests for this new functionality?

justinschuldt commented:
I'm going to kick this back into WIP while I write some tests

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+
+			for i := range sub.filters {
+				// get the filter by index/GetFilter rather than from range,
+				// so as to not copy the mutex within the Filter.
```
tl;dr - I fixed this by changing the `filters` struct to be an array of pointers, rather than an array of values.

`sub.filters` was `[]spyv1.FilterEntry`. When assigning the values in that range, ie `filterEntry` in the example below:
```golang
for _, filterEntry := range sub.filters {
  // ...
}
```

the values were being copied, which was throwing: 
> range var filterEntry copies lock:
> github.com/certusone/wormhole/node/pkg/proto/spy/v1.FilterEntry contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex `copylocks`


I changed `filters` to be references, rather than values, and updated this. Good question, thanks!

bruce-riley commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+	var topRes *spyv1.SubscribeSignedVAAByTypeResponse
+
+	// if v has values, the VAA unmarshal was successful.
+	if len(v.Payload) > 0 {
```
Can't v be nil?

justinschuldt replied:
I refactored this function into separate functions for the different VAA types, it's much cleaner, and it gets rid this hacky "is struct populated" check.

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+	// the unmarshaling could populate one of the variable below. it's also possible
+	// a new type of vaaBytes comes through, so the variables below could remain empty.
+
+	var v *vaa.VAA
```
refactored, this is gone

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+			if obsVAA.EmitterChain == vaa.ChainID(f.ChainId) {
+				// the emitter chain of the observation matches the filter
+
+				if f.Nonce >= 1 {
```
Thanks, I changed the signature of this function and simplified it a fair bit.

I left the `if f.Nonce >= 1`, context is, messages from the same transaction with the same nonce are batched together, with the exception of messages with nonce: `0`. It's a special case to allow developers to exclude messages from a batch. So because `0` has a special meaning, I opted to leave it `>= 1`

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+						v = b.Observations[0].Observation
+					}
+
+					if len(v.EmitterAddress) > 0 && v.EmitterChain > 0 {
```
Yeah, this was sketchy :see_no_evil:

Updated, thanks.

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+	}
+
+	// if b has vaules, the BatchVAA unmarshal was successful.
+	if len(b.Observations) > 0 {
```
fixed via refactor

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+		return matches, err
+	}
+
+	if txMatch {
```
done, simplified this a fair bit

justinschuldt commented on proto/gossip/v1/gossip.proto:
```diff
@@ -139,8 +139,10 @@ message SignedBatchObservation {
+  // Nonce of the messages in the batch.
+  uint32 nonce = 6;
+  // Batch ID - emitterChain/transactionID/nonce
+  string batch_id = 7;
```
Appreciate the flag here. There are no clients of BatchVaa protos yet, the first one will be this Spy code, and then the upcoming guardian changes that depend on this PR next.

The same applies to the other proto changes in this PR.

justinschuldt commented:
- I updated some protos (unused elsewhere) to align the `FilterEntry` types with the data they need to be compared against.
- I also updated the response types of the new Spy endpoint I added so we can return the `gossip` structs received by Spy, to make this code is a lot simpler, cleaner, etc.
- I added a bunch of tests for all the new functionality.
- The original entrypoint can/will be depreciated someday, the new functionality does everything it can do, and is more flexible, etc.

jynnantonix reviewed (dismissed):
Just some minor comments but otherwise looks great.  Thanks for adding tests!

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,17 +126,182 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+						sub.ch <- envelope
+					}
+				default:
+					return status.Error(codes.InvalidArgument, "unsupported filter type")
```
You're right, good idea, will change

justinschuldt commented on node/cmd/spy/spy_test.go:
```diff
@@ -0,0 +1,755 @@
+			}
+		}
+	}()
+	pause()
```
That's a fine idea, I'm happy to implement.

justinschuldt replied:
I changed this to a function that keeps checking for subscriptions and returns when it has one, no sleeping involved.

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,17 +126,182 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+						sub.ch <- envelope
+					}
+				default:
+					return status.Error(codes.InvalidArgument, "unsupported filter type")
```
updated to panic

justinschuldt commented on node/cmd/spy/spy_test.go:
```diff
@@ -0,0 +1,755 @@
+	msg1 := getBatchVAAQuorumMessage(vaa.ChainIDSolana, txID, nonce, emitter)
+	err = mockedSpyServer.HandleGossipBatchVAA(msg1)
+	if err != nil {
+		t.Log("failed HandleGossipBatchVAA for msg1")
+		t.FailNow()
```
Good eye, updated.

justinschuldt commented on node/cmd/spy/spy.go:
```diff
@@ -136,17 +126,182 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+
+	// loop through the subscriptions and send responses to everyone that wants this VAA
+	for _, sub := range s.subsAllVaa {
+		if len(sub.filters) == 0 {
+			// this subscription has no filters, send them the VAA.
+			sub.ch <- envelope
+		} else {
+			// this subscription has filters.
+			for _, filterEntry := range sub.filters {
```
Simpler is better, added this change.

jynnantonix commented on node/cmd/spy/spy.go:
```diff
@@ -69,84 +69,244 @@ var SpyCmd = &cobra.Command{
+					sub.ch <- envelope
 				}
+			default:
+				panic("unsupported filter type in subscriptions")
```
might also want to consider including the type here (via %T) to help with debugging:

```suggestion
				panic("unsupported filter type in subscriptions: %T", filter)
```

justinschuldt replied:
huh, TIL. added it, thanks.

jynnantonix commented on node/cmd/spy/spy_test.go:
```diff
@@ -0,0 +1,754 @@
+	if subs > 0 {
+		return
+	}
+	waitForClientSubscriptionInit(server)
```
Everything I've been able to find suggests that the go compiler doesn't implement recursive tail call optimization so this could overflow the stack.  We should stick to a loop.

Also we should add a `time.Sleep(time.Millisecond)` here to avoid unnecessary contention on the Mutex.

justinschuldt replied:
Interesting, thanks for bringing that up. I read a little about tail-call optimization, I believe you are correct; loops are the way. I updated this to use a loop with a 10 millisecond sleep.

jynnantonix reviewed (approved):
lgtm!

justinschuldt commented:
Closes https://github.com/wormhole-foundation/wormhole/issues/1891

justinschuldt commented:
Part of https://github.com/wormhole-foundation/wormhole/issues/1679
